File: src/main/java/me/anonymoussoftware/vacancymanager/ui/frames/MainFrame.java
package me.anonymoussoftware.vacancymanager.ui.frames;

import java.awt.BorderLayout;
import java.awt.Dimension;
import java.awt.EventQueue;
import java.awt.Toolkit;
import java.awt.event.ActionEvent;
import java.io.File;

import javax.swing.BoxLayout;
import javax.swing.JFileChooser;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JMenu;
import javax.swing.JMenuBar;
import javax.swing.JMenuItem;
import javax.swing.JOptionPane;
import javax.swing.JPanel;
import javax.swing.JProgressBar;
import javax.swing.JSplitPane;
import javax.swing.SwingConstants;
import javax.swing.border.BevelBorder;
import javax.swing.filechooser.FileNameExtensionFilter;

import me.anonymoussoftware.vacancymanager.App;
import me.anonymoussoftware.vacancymanager.VacancyManager;
import me.anonymoussoftware.vacancymanager.VacancyManager.VacancyDescriptionRequestListener;
import me.anonymoussoftware.vacancymanager.VacancyManager.VacancySearchListener;
import me.anonymoussoftware.vacancymanager.ui.panels.EmployerListPanel;
import me.anonymoussoftware.vacancymanager.ui.panels.SearchFormPanel;
import me.anonymoussoftware.vacancymanager.ui.panels.VacancyDescriptionPanel;
import me.anonymoussoftware.vacancymanager.ui.panels.VacancyListPanel;

@SuppressWarnings("serial")
public class MainFrame extends JFrame implements VacancySearchListener, VacancyDescriptionRequestListener {

    private static final int FRAME_WIDTH = 900;

    private static final int FRAME_HEIGHT = 700;

    private final VacancyListPanel vacancyListPanel;

    private final EmployerListPanel bannedEmployerListPanel;

    private final JLabel statusBarLabel;

    private final JProgressBar progressBar;

    private final VacancyManager vacancyManager = App.getBean(VacancyManager.class);

    public MainFrame() {
        setTitle("Vacancy manager");
        setSize(FRAME_WIDTH, FRAME_HEIGHT);
        Dimension screenSize = Toolkit.getDefaultToolkit().getScreenSize();
        setLocation((screenSize.width - FRAME_WIDTH) / 2, (screenSize.height - FRAME_HEIGHT) / 2);
        setLayout(new BorderLayout());

        add(new SearchFormPanel(), BorderLayout.NORTH);

        this.vacancyListPanel = new VacancyListPanel();
        this.bannedEmployerListPanel = new EmployerListPanel();
        JSplitPane innerSplitPane = new JSplitPane(JSplitPane.VERTICAL_SPLIT, this.vacancyListPanel,
                this.bannedEmployerListPanel);
        JSplitPane splitPane = new JSplitPane(JSplitPane.HORIZONTAL_SPLIT, innerSplitPane,
                new VacancyDescriptionPanel());
        add(splitPane, BorderLayout.CENTER);

        JPanel statusBarPanel = new JPanel();
        statusBarPanel.setBorder(new BevelBorder(BevelBorder.LOWERED));
        statusBarPanel.setPreferredSize(new Dimension(getWidth(), 20));
        statusBarPanel.setLayout(new BoxLayout(statusBarPanel, BoxLayout.X_AXIS));
        this.statusBarLabel = new JLabel("Status");
        this.statusBarLabel.setHorizontalAlignment(SwingConstants.LEFT);
        statusBarPanel.add(this.statusBarLabel);
        this.progressBar = new JProgressBar(0, 100);
        this.progressBar.setStringPainted(true);
        this.progressBar.setVisible(false);
        statusBarPanel.add(this.progressBar);
        add(statusBarPanel, BorderLayout.SOUTH);

        JMenuBar menubar = new JMenuBar();
        JMenu fileMenu = new JMenu("File");

        JMenu vacanciesMenu = new JMenu("Vacancies");
        JMenuItem importVacanciesMenuItem = new JMenuItem("Export...");
        importVacanciesMenuItem.addActionListener(this::onImportVacanciesMenuItemAction);
        vacanciesMenu.add(importVacanciesMenuItem);

        JMenuItem exportVacanciesMenuItem = new JMenuItem("Import...");
        exportVacanciesMenuItem.addActionListener(this::onExportVacanciesMenuItemAction);
        vacanciesMenu.add(exportVacanciesMenuItem);
        fileMenu.add(vacanciesMenu);

        JMenu employersMenu = new JMenu("Employers");
        JMenuItem importEmployersMenuItem = new JMenuItem("Export...");
        importEmployersMenuItem.addActionListener(this::onImportEmployersMenuItemAction);
        employersMenu.add(importEmployersMenuItem);

        JMenuItem exportEmployersMenuItem = new JMenuItem("Import...");
        exportEmployersMenuItem.addActionListener(this::onExportEmployersMenuItemAction);
        employersMenu.add(exportEmployersMenuItem);
        fileMenu.add(employersMenu);

        menubar.add(fileMenu);
        setJMenuBar(menubar);
    }

    @Override
    public void addNotify() {
        super.addNotify();
        this.vacancyManager.addVacancySearchListener(this);
        this.vacancyManager.addVacancyDescriptionRequestListener(this);
    }

    @Override
    public void removeNotify() {
        this.vacancyManager.removeVacancyDescriptionRequestListener(this);
        this.vacancyManager.removeVacancySearchListener(this);
        super.removeNotify();
    }

    private JFileChooser createFileChooser() {
        JFileChooser chooser = new JFileChooser();
        chooser.setCurrentDirectory(new File("."));
        chooser.setFileFilter(new FileNameExtensionFilter("JSON files", "json"));
        chooser.setAcceptAllFileFilterUsed(false);
        return chooser;
    }

    private void onImportVacanciesMenuItemAction(ActionEvent e) {
        JFileChooser chooser = createFileChooser();
        int result = chooser.showSaveDialog(this);
        if (result == JFileChooser.APPROVE_OPTION) {
            File selectedFile = chooser.getSelectedFile();
            String selectedFilePath = selectedFile.getPath();
            if (!selectedFilePath.toString().endsWith(".json")) {
                selectedFile = new File(selectedFilePath + ".json");
            }
            this.vacancyManager.importVacancies(selectedFile, //
                    this::showSuccessfullyLoadedVacanciesMessageDialog, //
                    this::showUnableToSaveVacanciesToFile);
        }
    }

    private void onExportVacanciesMenuItemAction(ActionEvent e) {
        JFileChooser chooser = createFileChooser();
        int result = chooser.showOpenDialog(this);
        if (result == JFileChooser.APPROVE_OPTION) {
            this.vacancyManager.exportVacancies(chooser.getSelectedFile(), //
                    this::showUnableToLoadVacanciesFromFileMessageDialog);
        }
    }

    private void onImportEmployersMenuItemAction(ActionEvent e) {
        JFileChooser chooser = createFileChooser();
        int result = chooser.showSaveDialog(this);
        if (result == JFileChooser.APPROVE_OPTION) {
            File selectedFile = chooser.getSelectedFile();
            String selectedFilePath = selectedFile.getPath();
            if (!selectedFilePath.toString().endsWith(".json")) {
                selectedFile = new File(selectedFilePath + ".json");
            }
            this.vacancyManager.importEmployers(selectedFile, //
                    this::showSuccessfullyLoadedEmployersMessageDialog, //
                    this::showUnableToSaveEmployersToFile);
        }
    }

    private void onExportEmployersMenuItemAction(ActionEvent e) {
        JFileChooser chooser = createFileChooser();
        int result = chooser.showOpenDialog(this);
        if (result == JFileChooser.APPROVE_OPTION) {
            this.vacancyManager.exportEmployers(chooser.getSelectedFile(), //
                    this::showUnableToLoadEmployersFromFileMessageDialog);
        }
    }

    private void showUnableToLoadVacanciesFromFileMessageDialog() {
        JOptionPane.showMessageDialog(this, "Unable to load vacancies from file", "Error", JOptionPane.ERROR_MESSAGE);
    }

    private void showSuccessfullyLoadedVacanciesMessageDialog() {
        JOptionPane.showMessageDialog(this, "Vacancies are saved successfully", "Success",
                JOptionPane.INFORMATION_MESSAGE);
    }

    private void showUnableToSaveVacanciesToFile() {
        JOptionPane.showMessageDialog(this, "Unable to save employers to file", "Error", JOptionPane.ERROR_MESSAGE);
    }

    private void showUnableToLoadEmployersFromFileMessageDialog() {
        JOptionPane.showMessageDialog(this, "Unable to load employers from file", "Error", JOptionPane.ERROR_MESSAGE);
    }

    private void showSuccessfullyLoadedEmployersMessageDialog() {
        JOptionPane.showMessageDialog(this, "Employers are saved successfully", "Success",
                JOptionPane.INFORMATION_MESSAGE);
    }

    private void showUnableToSaveEmployersToFile() {
        JOptionPane.showMessageDialog(this, "Unable to save vacancies to file", "Error", JOptionPane.ERROR_MESSAGE);
    }

    @Override
    public void onVacancyPageSuccessfulLoad(int percentage) {
        EventQueue.invokeLater(() -> {
            this.progressBar.setValue(percentage);
            this.progressBar.setString("Progress: " + percentage + "%");
            boolean isSearchCompleted = percentage == 100;
            if (isSearchCompleted) {
                this.statusBarLabel.setText("Vacancies are downloaded successfully");
            }
            this.statusBarLabel.setVisible(isSearchCompleted);
            this.progressBar.setVisible(!isSearchCompleted);
        });
    }

    @Override
    public void onStartVacancyPageLoad(String url) {
        EventQueue.invokeLater(
                () -> this.progressBar.setString(this.progressBar.getString() + " [Requesting " + url + "]"));
    }

    @Override
    public void onSearchFinish(int loaded, int total) {
        if (loaded == total) {
            JOptionPane.showMessageDialog(this, "Vacancy search is successfully done. Loaded: " + loaded, "Success",
                    JOptionPane.INFORMATION_MESSAGE);
        } else {
            JOptionPane.showMessageDialog(this,
                    "Vacancy search is done. Loaded " + loaded + " of " + total + " due to site restriction", "Warning",
                    JOptionPane.WARNING_MESSAGE);
        }

    }

    @Override
    public void onVacancyDescriptionRequestStart(String url) {
        EventQueue.invokeLater(() -> {
            this.progressBar.setString(this.progressBar.getString() + " [Requesting " + url + "]");
            this.statusBarLabel.setVisible(false);
            this.progressBar.setVisible(true);
        });
    }

    @Override
    public void onVacancyDescriptionRequestFinish() {
        EventQueue.invokeLater(() -> {
            this.statusBarLabel.setText("Vacancy description is downloaded successfully");
            this.statusBarLabel.setVisible(true);
            this.progressBar.setVisible(false);
        });
    }
}

File: src/main/kotlin/me/anonymoussoftware/vacancymanager/ui/frames/MainFrame.kt
package me.anonymoussoftware.vacancymanager.ui.frames

import me.anonymoussoftware.vacancymanager.App
import me.anonymoussoftware.vacancymanager.VacancyManager
import me.anonymoussoftware.vacancymanager.VacancyManager.VacancyDescriptionRequestListener
import me.anonymoussoftware.vacancymanager.VacancyManager.VacancySearchListener
import me.anonymoussoftware.vacancymanager.ui.panels.EmployerListPanel
import me.anonymoussoftware.vacancymanager.ui.panels.SearchFormPanel
import me.anonymoussoftware.vacancymanager.ui.panels.VacancyDescriptionPanel
import me.anonymoussoftware.vacancymanager.ui.panels.VacancyListPanel
import java.awt.BorderLayout
import java.awt.Dimension
import java.awt.EventQueue
import java.awt.Toolkit
import java.io.File
import javax.swing.*
import javax.swing.border.BevelBorder
import javax.swing.filechooser.FileNameExtensionFilter

class MainFrame : JFrame(), VacancySearchListener, VacancyDescriptionRequestListener {

    private val vacancyListPanel: VacancyListPanel

    private val bannedEmployerListPanel: EmployerListPanel

    private val statusBarLabel: JLabel

    private val progressBar: JProgressBar

    private val vacancyManager = App.getBean(VacancyManager::class.java)

    init {
        title = "Vacancy manager"
        setSize(FRAME_WIDTH, FRAME_HEIGHT)
        val screenSize = Toolkit.getDefaultToolkit().screenSize
        setLocation((screenSize.width - FRAME_WIDTH) / 2, (screenSize.height - FRAME_HEIGHT) / 2)
        layout = BorderLayout()

        add(SearchFormPanel(), BorderLayout.NORTH)

        this.vacancyListPanel = VacancyListPanel()
        this.bannedEmployerListPanel = EmployerListPanel()
        val innerSplitPane = JSplitPane(
            JSplitPane.VERTICAL_SPLIT, this.vacancyListPanel,
            this.bannedEmployerListPanel
        )
        val splitPane = JSplitPane(
            JSplitPane.HORIZONTAL_SPLIT, innerSplitPane,
            VacancyDescriptionPanel()
        )
        add(splitPane, BorderLayout.CENTER)

        val statusBarPanel = JPanel()
        statusBarPanel.border = BevelBorder(BevelBorder.LOWERED)
        statusBarPanel.preferredSize = Dimension(width, 20)
        statusBarPanel.layout = BoxLayout(statusBarPanel, BoxLayout.X_AXIS)
        this.statusBarLabel = JLabel("Status")
        this.statusBarLabel.horizontalAlignment = SwingConstants.LEFT
        statusBarPanel.add(this.statusBarLabel)
        this.progressBar = JProgressBar(0, 100)
        this.progressBar.isStringPainted = true
        this.progressBar.isVisible = false
        statusBarPanel.add(this.progressBar)
        add(statusBarPanel, BorderLayout.SOUTH)

        val menubar = JMenuBar()
        val fileMenu = JMenu("File")

        val vacanciesMenu = JMenu("Vacancies")
        val importVacanciesMenuItem = JMenuItem("Export...")
        importVacanciesMenuItem.addActionListener( { this.onImportVacanciesMenuItemAction() })
        vacanciesMenu.add(importVacanciesMenuItem)

        val exportVacanciesMenuItem = JMenuItem("Import...")
        exportVacanciesMenuItem.addActionListener( { this.onExportVacanciesMenuItemAction() })
        vacanciesMenu.add(exportVacanciesMenuItem)
        fileMenu.add(vacanciesMenu)

        val employersMenu = JMenu("Employers")
        val importEmployersMenuItem = JMenuItem("Export...")
        importEmployersMenuItem.addActionListener( { this.onImportEmployersMenuItemAction() })
        employersMenu.add(importEmployersMenuItem)

        val exportEmployersMenuItem = JMenuItem("Import...")
        exportEmployersMenuItem.addActionListener { this.onExportEmployersMenuItemAction() }
        employersMenu.add(exportEmployersMenuItem)
        fileMenu.add(employersMenu)

        menubar.add(fileMenu)
        jMenuBar = menubar
    }

    override fun addNotify() {
        super.addNotify()
        this.vacancyManager.addVacancySearchListener(this)
        this.vacancyManager.addVacancyDescriptionRequestListener(this)
    }

    override fun removeNotify() {
        this.vacancyManager.removeVacancyDescriptionRequestListener(this)
        this.vacancyManager.removeVacancySearchListener(this)
        super.removeNotify()
    }

    private fun createFileChooser(): JFileChooser {
        val chooser = JFileChooser()
        chooser.currentDirectory = File(".")
        chooser.fileFilter = FileNameExtensionFilter("JSON files", "json")
        chooser.isAcceptAllFileFilterUsed = false
        return chooser
    }

    private fun onImportVacanciesMenuItemAction() {
        val chooser = createFileChooser()
        val result = chooser.showSaveDialog(this)
        if (result == JFileChooser.APPROVE_OPTION) {
            var selectedFile = chooser.selectedFile
            val selectedFilePath = selectedFile.path
            if (!selectedFilePath.toString().endsWith(".json")) {
                selectedFile = File("$selectedFilePath.json")
            }
            this.vacancyManager.importVacancies(selectedFile, //
                Runnable { this.showSuccessfullyLoadedVacanciesMessageDialog() }, //
                Runnable { this.showUnableToSaveVacanciesToFile() })
        }
    }

    private fun onExportVacanciesMenuItemAction() {
        val chooser = createFileChooser()
        val result = chooser.showOpenDialog(this)
        if (result == JFileChooser.APPROVE_OPTION) {
            this.vacancyManager.exportVacancies(chooser.selectedFile, //
                Runnable { this.showUnableToLoadVacanciesFromFileMessageDialog() })
        }
    }

    private fun onImportEmployersMenuItemAction() {
        val chooser = createFileChooser()
        val result = chooser.showSaveDialog(this)
        if (result == JFileChooser.APPROVE_OPTION) {
            var selectedFile = chooser.selectedFile
            val selectedFilePath = selectedFile.path
            if (!selectedFilePath.toString().endsWith(".json")) {
                selectedFile = File("$selectedFilePath.json")
            }
            this.vacancyManager.importEmployers(selectedFile, //
                Runnable { this.showSuccessfullyLoadedEmployersMessageDialog() }, //
                Runnable { this.showUnableToSaveEmployersToFile() })
        }
    }

    private fun onExportEmployersMenuItemAction() {
        val chooser = createFileChooser()
        val result = chooser.showOpenDialog(this)
        if (result == JFileChooser.APPROVE_OPTION) {
            this.vacancyManager.exportEmployers(chooser.selectedFile, //
                Runnable { this.showUnableToLoadEmployersFromFileMessageDialog() })
        }
    }

    private fun showUnableToLoadVacanciesFromFileMessageDialog() {
        JOptionPane.showMessageDialog(this, "Unable to load vacancies from file", "Error", JOptionPane.ERROR_MESSAGE)
    }

    private fun showSuccessfullyLoadedVacanciesMessageDialog() {
        JOptionPane.showMessageDialog(
            this, "Vacancies are saved successfully", "Success",
            JOptionPane.INFORMATION_MESSAGE
        )
    }

    private fun showUnableToSaveVacanciesToFile() {
        JOptionPane.showMessageDialog(this, "Unable to save employers to file", "Error", JOptionPane.ERROR_MESSAGE)
    }

    private fun showUnableToLoadEmployersFromFileMessageDialog() {
        JOptionPane.showMessageDialog(this, "Unable to load employers from file", "Error", JOptionPane.ERROR_MESSAGE)
    }

    private fun showSuccessfullyLoadedEmployersMessageDialog() {
        JOptionPane.showMessageDialog(
            this, "Employers are saved successfully", "Success",
            JOptionPane.INFORMATION_MESSAGE
        )
    }

    private fun showUnableToSaveEmployersToFile() {
        JOptionPane.showMessageDialog(this, "Unable to save vacancies to file", "Error", JOptionPane.ERROR_MESSAGE)
    }

    override fun onVacancyPageSuccessfulLoad(percentage: Int) {
        EventQueue.invokeLater {
            this.progressBar.value = percentage
            this.progressBar.string = "Progress: $percentage%"
            val isSearchCompleted = percentage == 100
            if (isSearchCompleted) {
                this.statusBarLabel.text = "Vacancies are downloaded successfully"
            }
            this.statusBarLabel.isVisible = isSearchCompleted
            this.progressBar.isVisible = !isSearchCompleted
        }
    }

    override fun onStartVacancyPageLoad(url: String) {
        EventQueue.invokeLater { this.progressBar.string = this.progressBar.string + " [Requesting " + url + "]" }
    }

    override fun onSearchFinish(loaded: Int, total: Int) {
        if (loaded == total) {
            JOptionPane.showMessageDialog(
                this, "Vacancy search is successfully done. Loaded: $loaded", "Success",
                JOptionPane.INFORMATION_MESSAGE
            )
        } else {
            JOptionPane.showMessageDialog(
                this,
                "Vacancy search is done. Loaded $loaded of $total due to site restriction", "Warning",
                JOptionPane.WARNING_MESSAGE
            )
        }

    }

    override fun onVacancyDescriptionRequestStart(url: String) {
        EventQueue.invokeLater {
            this.progressBar.string = this.progressBar.string + " [Requesting " + url + "]"
            this.statusBarLabel.isVisible = false
            this.progressBar.isVisible = true
        }
    }

    override fun onVacancyDescriptionRequestFinish() {
        EventQueue.invokeLater {
            this.statusBarLabel.text = "Vacancy description is downloaded successfully"
            this.statusBarLabel.isVisible = true
            this.progressBar.isVisible = false
        }
    }

    companion object {

        private val FRAME_WIDTH = 900

        private val FRAME_HEIGHT = 700
    }
}
